ereuse_devicehub/resources/device/views.py

```diff
@@ -152,6 +156,67 @@ def query(self, args):
         return query.filter(*args['filter']).order_by(*args['sort'])
 
 
+class DeviceMergeView(View):
+
+    """View for merging two devices
+    Ex. ``device/<id>/merge/id=X``.
+    """
+    class FindArgs(MarshmallowSchema):
+        id = fields.Integer()
+
+    def get_merge_id(self) -> uuid.UUID:
+        args = self.QUERY_PARSER.parse(self.find_args, request, locations=('querystring',))
+        return args['id']
+
+    def post(self, id: uuid.UUID):
+        device = Device.query.filter_by(id=id).one()
+        with_device = Device.query.filter_by(id=self.get_merge_id()).one()
+        self.merge_devices(device, with_device)
+
+        db.session().final_flush()
+        ret = self.schema.jsonify(device)
+        ret.status_code = 201
+
+        db.session.commit()
+        return ret
+
+    def merge_devices(self, base_device, with_device):
+        """Merge the current device with `with_device` by
+        adding all `with_device` actions under the current device.
+
+        This operation is highly costly as it forces refreshing
+        many models in session.
+        """
+        snapshots = sorted(filterfalse(lambda x: not isinstance(x, actions.Snapshot), (base_device.actions + with_device.actions)))
+        workbench_snapshots = [ s for s in snapshots if s.software == (SnapshotSoftware.Workbench or SnapshotSoftware.WorkbenchAndroid)]
+        latest_snapshot_device = [ d for d in (base_device, with_device) if d.id == snapshots[-1].device.id][0]
+        latest_snapshotworkbench_device = [ d for d in (base_device, with_device) if d.id == workbench_snapshots[-1].device.id][0]
+        # Adding actions of with_device
+        with_actions_one = [a for a in with_device.actions if isinstance(a, actions.ActionWithOneDevice)]
+        with_actions_multiple = [a for a in with_device.actions if isinstance(a, actions.ActionWithMultipleDevices)]
```

Comment:
why do you split up actions in `with_actions_one` and `with_actions_multiple`?

Reply:
no, because if you see at line 142 on [device/models.py](https://github.com/eReuse/devicehub-teal/blob/9ad217fc5ba058f577f86ffa3b5f191056132c88/ereuse_devicehub/resources/device/models.py#L142) actions is a virtual property accumulating `actions_one` and `actions_multiple`